Reusable kubernetes angularjs client

stefwalter opened the pull request:
This is the implementation of a reusable angularjs client. The Cockpit specific stuff (such as how to do a watch, and how to do a request) is in its own pluggable module.

stefwalter commented:
Did some documentation, cleanup, and brought over methods. Almost certainly doesn't work yet. Next is testing.

liggitt commented:
@jwforres fyi

sub-mod commented on pkg/registry/scripts/kube-client.js:
```diff
@@ -0,0 +1,1033 @@
+     *  2. a path string
+     *  3. type/kind, name, namespace
+     */
+    function resource_path(args) {
```
is args is a string this fails.for ex: resource_path("images") will fail

stefwalter replied:
Is there a case where it's called like this? I think all the callers pass an array of arguments.

sub-mod commented on pkg/registry/scripts/images.js:
```diff
@@ -36,8 +38,15 @@
+        function($scope, loader, select) {
+
             /* nothing here yet */
+            loader.watch("images");
```
should this be loader.watch(]"images"]); ?

stefwalter replied:
I don't think that's how ```loader.watch()``` accepts its arguments. See the documentation in comments near kubeLoader.

sub-mod commented on pkg/registry/scripts/kube-client.js:
```diff
@@ -0,0 +1,1033 @@
+    var KUBE = "/api/v1";
+    var OPENSHIFT = "/oapi/v1";
+    var DEFAULT = { api: KUBE, create: 0 };
+    var SCHEMA = flat_schema([
```
do we add info on which has watch and which doesnt ?

stefwalter replied:
Sure. But what would we do with such info? It would seem that the caller would need to special case these anyway and load entries completely differently ... using ```loader.load()``` instead of ```loader.watch()```

Can you describe an example?

stefwalter replied:
Lets add this info later as a separate pull request, once we have something that consumes it.

sub-mod commented on pkg/registry/scripts/kube-client.js:
```diff
@@ -0,0 +1,1033 @@
+        { kind: "Pod", type: "pods", api: KUBE, create: -20 },
+        { kind: "ReplicationController", type: "replicationcontrollers", api: KUBE, create: -60 },
+        { kind: "Service", type: "services", api: KUBE, create: -80 },
+        { kind: "User", type: "user", api: OPENSHIFT, global: true },
```
should be users
{ kind: "User", type: "users", api: OPENSHIFT, global: true },

stefwalter replied:
Will do.

sub-mod commented on pkg/registry/scripts/kube-client-cockpit.js:
```diff
@@ -0,0 +1,652 @@
+                 * connect discovery stuff. Otherwise ask our connect service
+                 * for connection info, and do discovery.
+                 */
+                if (config.payload)
```
I think this should be 
if (config && config.payload)

stefwalter replied:
Good catch.

sub-mod commented on pkg/registry/scripts/kube-client.js:
```diff
@@ -0,0 +1,1033 @@
+        { kind: "Group", type: "groups", api: OPENSHIFT },
+        { kind: "Image", type: "images", api: OPENSHIFT, global: true },
+        { kind: "ImageStreams", type: "imagestreams", api: OPENSHIFT },
+        { kind: "Namespace", type: "namespaces", api: KUBE, global: true, create: -100 },
```
What should be the create value for Projects ?
{ kind: "Project", type: "projects", api: OPENSHIFT, global: true, create: -90    } is this ok ?

stefwalter replied:
Yes, that seems right. Will add.

sub-mod commented:
I can see an error 
Error: [$rootScope:inprog] $digest already in progress
http://errors.angularjs.org/1.3.14/$rootScope/inprog?p0=%24digest
    at REGEX_STRING_REGEXP (angular.js:65)
    at beginPhase (angular.js:14757)
    at Scope.$get.Scope.$digest (angular.js:14199)
    at Object.<anonymous> (main.js:64)
    at invoke_callbacks (kube-client.js:361)
    at handle_flush (kube-client.js:352)
    at handle_frames (kube-client.js:322)
    at handle_objects (kube-client.js:408)
    at angular.module.value.factory.req.then.req (kube-client.js:478)
    at processQueue (angular.js:13191)

stefwalter commented:
> I can see an error 

 * Fixed this by using applyAsync

Also:

 * Made loader.watch() accept same sort of arguments as loader.load()
 * Removed unused loader pull_queue() related logic.
 * Fixed promise returned by CockpitKubeWatch
 * Fixed bugs in loader.load() processing of objects
 * Added ImageStreamImage to schema

stefwalter commented:
My latest changes broke namespaced watches. Test is also false negative.

stefwalter commented:
Fixed the above problem, and also used a javascript naming style more similar to angular and openshift.

sub-mod commented on pkg/registry/scripts/kube-client.js:
```diff
@@ -0,0 +1,985 @@
+            var everything = null;
+
+            loader.listen(function(present, removed) {
+console.log("listen");
```
indentation ?

stefwalter replied:
We should  remove this line.

sub-mod commented:
LGTM

petervo commented:
I noticed that the tests weren't included in the makefile so i did 

```
diff --git a/pkg/registry/Makefile.am b/pkg/registry/Makefile.am
index d22fbe8..b9b919a 100644
--- a/pkg/registry/Makefile.am
+++ b/pkg/registry/Makefile.am
@@ -25,6 +25,7 @@ registry_BOWER = \
        $(NULL)
 
 registry_TESTS = \
+    pkg/registry/tests/test-kube-client.html \
        $(NULL)
 
 # Typically only modify stuff above this line
@@ -98,6 +99,8 @@ EXTRA_DIST += \
        $(registrydebug_DATA) \
        $(registry_TESTS) \
        $(registry_VIEWS) \
+    pkg/registry/tests/fixture-basic.js \
+    pkg/registry/tests/fixture-large.js \
        $(NULL)
```

With that the tests run with make check, but i'm getting failures.

FAIL: pkg/registry/tests/test-kube-client.html 25 - pods gone, expected: 0, got: 3, test: set namespace, source: at http://localhost:41576/tools/qunit.js:10753
FAIL: pkg/registry/tests/test-kube-client.html 26 - pods from namespace other, expected: 1, got: 3, test: set namespace, source: at http://localhost:41576/tools/qunit.js:10753

stefwalter commented:
Thanks. Good catch, will fix.

stefwalter commented:
Included the tests, and fixed the bugs found. An extra fix needed for phantomjs, but this should make us more robust in the end. Removed an unused variable as well.

stefwalter commented:
Fixed 'make distcheck' issue.
